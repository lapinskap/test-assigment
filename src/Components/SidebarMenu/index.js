import PropTypes from 'prop-types';
import { Link, useHistory } from 'react-router-dom';
import React, { Fragment, useEffect, useState } from 'react';
import __ from '../../utils/Translations';

export default function SidebarMenu({ content, linkId, isSubMenu }) {
  const history = useHistory();
  const [activeId, setActiveId] = useState(
    linkId || getActiveId(content, history.location.pathname),
  );
  const [openedSubmenu, setOpenedSubMenu] = useState(getDefaultOpenedTab(content, activeId));
  const toggleSubmenu = (submenuId) => {
    setOpenedSubMenu(submenuId === openedSubmenu ? '' : submenuId);
  };

  useEffect(() => history.listen((location) => {
    const newId = getActiveId(content, location.pathname);
    setActiveId(newId);
    if (newId) {
      const item = content.find((el) => el.id === newId);
      if (item && item.content && item.content.length) {
        setOpenedSubMenu(newId);
      }
    }
  }));
  return (
    <div className={isSubMenu ? 'sub-menu metismenu vertical-nav-menu' : 'main-menu metismenu vertical-nav-menu'}>
      <ul className="metismenu-container" data-t1="sidebarMenu">
        {content.map((item) => {
          const {
            id, to, content: itemContent, label, icon,
          } = item;
          const hasSubmenu = itemContent && itemContent.length;
          const hasActiveChild = hasSubmenu && itemContent.find(
            (subItem) => subItem.id === activeId,
          );
          const isOpen = hasSubmenu && (openedSubmenu === id);
          const isActive = activeId === id;
          return (
            <Fragment key={id}>
              <li className="metismenu-item">
                <Link
                  data-t1={id}
                  className={`metismenu-link${hasActiveChild ? ' has-active-child' : ''}${isActive ? ' active' : ''}`}
                  to={to || '#'}
                  onClick={() => toggleSubmenu(id)}
                >
                  <i className={`metismenu-icon${icon ? ` ${icon}` : ''}`} />
                  {__(label)}
                  {hasSubmenu ? (
                    <i className={`metismenu-state-icon pe-7s-angle-down caret-left${isOpen ? ' rotate-minus-90' : ''}`} />
                  ) : null}
                </Link>
                {hasSubmenu ? (
                  <ul className={`metismenu-container${isOpen ? ' visible' : ''}`}>
                    {itemContent.map((subItem) => (
                      <li className="metismenu-item" key={subItem.id}>
                        <Link
                          data-t1={subItem.id}
                          className={`metismenu-link${activeId === subItem.id ? ' active' : ''}`}
                          to={subItem.to}
                        >
                          <i className={`metismenu-icon${subItem.icon ? ` ${subItem.icon}` : ''}`} />
                          {__(subItem.label)}
                        </Link>
                      </li>
                    ))}
                  </ul>
                ) : null}
              </li>
            </Fragment>
          );
        })}
      </ul>
    </div>
  );
}

const getDefaultOpenedTab = (content, activeId) => {
  const tab = content.find(
    (item) => item.content && item.content.find((subItem) => subItem.id === activeId),
  );
  return tab ? tab.id : null;
};

export const isActiveItem = (item, pathname, isAnchor) => {
  if (item.to) {
    return isAnchor ? pathname.startsWith(item.to) : item.to === pathname;
  }
  return false;
};

export const getActiveId = (content, pathname) => {
  for (let i = 0; i < content.length; i += 1) {
    const item = content[i];
    if (isActiveItem(item, pathname, item.isAnchor)) {
      return item.id;
    }
    if (item.content) {
      const subItem = item.content.find((el) => isActiveItem(el, pathname, true));
      if (subItem) {
        return subItem.id;
      }
    }
  }
  return null;
};

SidebarMenu.propTypes = {
  content: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    to: PropTypes.string,
    label: PropTypes.string,
    icon: PropTypes.string,
    strict: PropTypes.bool,
    content: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string,
      to: PropTypes.string,
      label: PropTypes.string,
      icon: PropTypes.string,
      strict: PropTypes.bool,
    })),
  })).isRequired,
  linkId: PropTypes.string,
  isSubMenu: PropTypes.bool.isRequired,
};

SidebarMenu.defaultProps = {
  linkId: '',
};
