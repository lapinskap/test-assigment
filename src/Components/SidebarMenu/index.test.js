/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/prop-types */
import React from 'react';
// import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import SidebarMenu, { isActiveItem, getActiveId } from './index';
// import TestAppContext from '../../tests/TestAppContext';
import '../../tests/setupTests';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    listen: () => ({ location: 'test' }),
  }),
  Link: ({ to, children, ...props }) => <a href={to} {...props}>{children}</a>,
}));

describe('isActiveItem()', () => {
  it('executes isActiveItem', () => {
    const item = [{ to: 'some', label: 'Some', content: [{ to: 'some employees', label: 'None' }] }];
    const pathname = '/';
    const result = isActiveItem(item, pathname);
    expect(result).toBe(false);
  });
});

describe('getActiveId', () => {
  it('executes getActiveId', () => {
    const content = [{
      id: 1, to: 'some', label: 'Some', content: [{ to: 'some employees', label: 'None' }],
    }];
    const pathname = '/';
    const result = getActiveId(content, pathname);
    expect(result).toBe(null);
  });
});

describe('<SidebarMenu />', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  let content;
  let
    linkId;

  beforeEach(() => {
    content = [{
      icon: 'pe-7s-diamond',
      label: 'Katalog produktów',
      to: '/product-catalog/products',
      id: 'product-catalog',
      aclKey: 'product-catalog',
    }, {
      icon: 'pe-7s-shuffle',
      label: 'Lista dostawców',
      to: '/suppliers',
      id: 'provider',
      aclKey: 'provider',
    }, {
      icon: 'pe-7s-id',
      label: 'Firma',
      id: 'company',
      aclKey: 'company',
      content: [{
        label: 'Lista firm',
        to: '/company/list',
        id: 'company_list',
        aclKey: 'company_list',
      }, {
        label: 'Pracownicy',
        to: '/company/employee-list',
        id: 'employee_list',
        aclKey: 'employee_list',
      }],
    }, {
      icon: 'pe-7s-tools',
      label: 'Administracja',
      id: 'administration',
      aclKey: 'administration',
      content: [{
        label: 'Parametry konfiguracyjne',
        to: '/administration/configuration/default/default',
        id: 'configuration',
        aclKey: 'configuration',
      }, {
        label: 'Słowniki',
        to: '/administration/dictionary',
        id: 'dictionary',
      }, {
        label: 'Konfiguracja adresów IP',
        to: '/administration/ip-security',
        id: 'ip-security',
        aclKey: 'ip-security',
      }],
    }, {
      icon: 'pe-7s-users',
      label: 'Operatorzy MB',
      to: '/user/operator/list',
      id: 'user_operator_list',
      aclKey: 'user_operator_list',
    }, {
      icon: 'pe-7s-box2',
      label: 'CMS',
      to: '/cms/management',
      id: 'cms',
      aclKey: 'cms',
    }, {
      icon: 'pe-7s-graph2',
      label: 'Raporty',
      id: 'report',
      aclKey: 'report',
      content: [{
        label: 'Wybrane benefity',
        to: '/report/chosen_subscription_benefits',
        id: 'chosen_subscription_benefits',
        aclKey: 'chosen_subscription_benefits',
      }],
    }, {
      icon: 'pe-7s-notebook',
      label: 'Tłumaczenia',
      id: 'translate',
      aclKey: 'report',
      content: [{
        label: 'Tłumaczenia interfejsu',
        to: '/translate/interface',
        id: 'translate_interface',
        aclKey: 'translate_interface',
      }, {
        label: 'Tłumaczenia wartości prostych',
        to: '/translate/simple',
        id: 'translate_simple',
        aclKey: 'translate_simple',
      }, {
        label: 'Tłumaczenia opisów',
        to: '/translate/description',
        id: 'translate_description',
        aclKey: 'translate_description',
      }, {
        label: 'Zakresy',
        to: '/translate/scopes',
        id: 'translate_scopes',
        aclKey: 'translate_scopes',
      }],
    }, {
      icon: 'pe-7s-display2',
      label: 'Standardy',
      id: 'standards',
      aclKey: 'standards',
      content: [{
        label: 'Grid filtrowalny js',
        to: '/standards/grid/filtrable-js',
        id: 'grid_filtrerable_js',
        aclKey: 'standards',
      }, {
        label: 'Grid filtrowalny backend',
        to: '/standards/grid/filtrable-backend',
        id: 'grid_filtrerable_backend',
        aclKey: 'standards',
      }, {
        label: 'Formularze',
        to: '/standards/forms',
        id: 'forms',
        aclKey: 'standards',
      }, {
        label: 'Zakładki (proste)',
        to: '/standards/tabs/simple',
        id: 'tabs_simple',
        aclKey: 'standards',
      }, {
        label: 'Zakładki (z zapamiętywaniem)',
        to: '/standards/tabs/with-memo',
        id: 'tabs_with_memo',
        aclKey: 'standards',
      }, {
        label: 'Zakładki boczne',
        to: '/standards/tabs/sidebar',
        id: 'tabs_sidebar',
        aclKey: 'standards',
      }],
    }];
    linkId = '12';
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default state', () => {
    const tree = renderer
      .create(<SidebarMenu content={content} linkId={linkId} />)
      .toJSON();

    expect(tree)
      .toMatchSnapshot();
  });

  // it('should call onClick on Link', async () => {
  //   const wrapper = mount(<TestAppContext router><SidebarMenu content={content} linkId={linkId} /></TestAppContext>);
  //
  //   const toggleLink = wrapper.find('Link').at(0);
  //   const toggleSpy = jest.fn();
  //   await act(async () => {
  //     await toggleLink.simulate('click');
  //   });
  //   wrapper.update();
  //   expect(await toggleSpy).toHaveBeenCalled();
  // });
});
