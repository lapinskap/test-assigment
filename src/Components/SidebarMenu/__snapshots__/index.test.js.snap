// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarMenu /> renders correctly with default state 1`] = `
<div
  className="main-menu metismenu vertical-nav-menu"
>
  <ul
    className="metismenu-container"
    data-t1="sidebarMenu"
  >
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="product-catalog"
        href="/product-catalog/products"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-diamond"
        />
        Katalog produktów
      </a>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="provider"
        href="/suppliers"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-shuffle"
        />
        Lista dostawców
      </a>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="company"
        href="#"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-id"
        />
        Firma
        <i
          className="metismenu-state-icon pe-7s-angle-down caret-left"
        />
      </a>
      <ul
        className="metismenu-container"
      >
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="company_list"
            href="/company/list"
          >
            <i
              className="metismenu-icon"
            />
            Lista firm
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="employee_list"
            href="/company/employee-list"
          >
            <i
              className="metismenu-icon"
            />
            Pracownicy
          </a>
        </li>
      </ul>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="administration"
        href="#"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-tools"
        />
        Administracja
        <i
          className="metismenu-state-icon pe-7s-angle-down caret-left"
        />
      </a>
      <ul
        className="metismenu-container"
      >
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="configuration"
            href="/administration/configuration/default/default"
          >
            <i
              className="metismenu-icon"
            />
            Parametry konfiguracyjne
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="dictionary"
            href="/administration/dictionary"
          >
            <i
              className="metismenu-icon"
            />
            Słowniki
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="ip-security"
            href="/administration/ip-security"
          >
            <i
              className="metismenu-icon"
            />
            Konfiguracja adresów IP
          </a>
        </li>
      </ul>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="user_operator_list"
        href="/user/operator/list"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-users"
        />
        Operatorzy MB
      </a>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="cms"
        href="/cms/management"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-box2"
        />
        CMS
      </a>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="report"
        href="#"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-graph2"
        />
        Raporty
        <i
          className="metismenu-state-icon pe-7s-angle-down caret-left"
        />
      </a>
      <ul
        className="metismenu-container"
      >
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="chosen_subscription_benefits"
            href="/report/chosen_subscription_benefits"
          >
            <i
              className="metismenu-icon"
            />
            Wybrane benefity
          </a>
        </li>
      </ul>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="translate"
        href="#"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-notebook"
        />
        Tłumaczenia
        <i
          className="metismenu-state-icon pe-7s-angle-down caret-left"
        />
      </a>
      <ul
        className="metismenu-container"
      >
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="translate_interface"
            href="/translate/interface"
          >
            <i
              className="metismenu-icon"
            />
            Tłumaczenia interfejsu
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="translate_simple"
            href="/translate/simple"
          >
            <i
              className="metismenu-icon"
            />
            Tłumaczenia wartości prostych
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="translate_description"
            href="/translate/description"
          >
            <i
              className="metismenu-icon"
            />
            Tłumaczenia opisów
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="translate_scopes"
            href="/translate/scopes"
          >
            <i
              className="metismenu-icon"
            />
            Zakresy
          </a>
        </li>
      </ul>
    </li>
    <li
      className="metismenu-item"
    >
      <a
        className="metismenu-link"
        data-t1="standards"
        href="#"
        onClick={[Function]}
      >
        <i
          className="metismenu-icon pe-7s-display2"
        />
        Standardy
        <i
          className="metismenu-state-icon pe-7s-angle-down caret-left"
        />
      </a>
      <ul
        className="metismenu-container"
      >
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="grid_filtrerable_js"
            href="/standards/grid/filtrable-js"
          >
            <i
              className="metismenu-icon"
            />
            Grid filtrowalny js
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="grid_filtrerable_backend"
            href="/standards/grid/filtrable-backend"
          >
            <i
              className="metismenu-icon"
            />
            Grid filtrowalny backend
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="forms"
            href="/standards/forms"
          >
            <i
              className="metismenu-icon"
            />
            Formularze
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="tabs_simple"
            href="/standards/tabs/simple"
          >
            <i
              className="metismenu-icon"
            />
            Zakładki (proste)
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="tabs_with_memo"
            href="/standards/tabs/with-memo"
          >
            <i
              className="metismenu-icon"
            />
            Zakładki (z zapamiętywaniem)
          </a>
        </li>
        <li
          className="metismenu-item"
        >
          <a
            className="metismenu-link"
            data-t1="tabs_sidebar"
            href="/standards/tabs/sidebar"
          >
            <i
              className="metismenu-icon"
            />
            Zakładki boczne
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
